transport/session/src/balancer/controller.rs

```diff
@@ -0,0 +1,298 @@
+    /// The default is 2500 (which is 1250 packets/second currently)
+    #[default(2500)]
+    pub max_surbs_per_sec: u64,
+}
+
+/// Runs a continuous process that attempts to [evaluate](SurbFlowEstimator) and
+/// [regulate](SurbFlowController) the flow of non-organic SURBs to the Session counterparty,
+/// to keep the number of SURBs at the counterparty at a certain level.
+///
+/// Internally, the Balancer uses a PID (Proportional Integral Derivative) controller to
+/// control the rate of SURBs sent to the counterparty
+/// each time the [`update`](SurbBalancer::update) method is called:
+///
+/// 1. The size of the SURB buffer at the counterparty is estimated using [`SurbFlowEstimator`].
+/// 2. Error against a set-point given in [`SurbBalancerConfig`] is evaluated in the PID controller.
+/// 3. The PID controller applies a new SURB flow rate value using the [`SurbFlowController`].
+pub struct SurbBalancer<I, O, F, S> {
+    session_id: S,
+    pid: Pid<f64>,
+    surb_production_estimator: O,
+    surb_consumption_estimator: I,
+    controller: F,
+    cfg: SurbBalancerConfig,
+    current_target_buffer: u64,
+    last_surbs_delivered: u64,
+    last_surbs_used: u64,
+    last_update: std::time::Instant,
+    was_below_target: bool,
+}
+
+// Default coefficients for the PID controller
+const DEFAULT_P_GAIN: f64 = 0.6;
+const DEFAULT_I_GAIN: f64 = 0.7;
+const DEFAULT_D_GAIN: f64 = 0.2;
+
+impl<I, O, F, S> SurbBalancer<I, O, F, S>
+where
+    O: SurbFlowEstimator,
+    I: SurbFlowEstimator,
+    F: SurbFlowController,
+    S: Display,
+{
+    pub fn new(
+        session_id: S,
+        surb_production_estimator: O,
+        surb_consumption_estimator: I,
+        controller: F,
+        cfg: SurbBalancerConfig,
+    ) -> Self {
+        // Initialize the PID controller with default tuned gains
+        let mut pid = Pid::new(cfg.target_surb_buffer_size as f64, cfg.max_surbs_per_sec as f64);
+        pid.p(DEFAULT_P_GAIN, cfg.max_surbs_per_sec as f64);
+        pid.i(DEFAULT_I_GAIN, cfg.max_surbs_per_sec as f64);
+        pid.d(DEFAULT_D_GAIN, cfg.max_surbs_per_sec as f64);
+
+        #[cfg(all(feature = "prometheus", not(test)))]
+        {
+            let sid = session_id.to_string();
+            METRIC_TARGET_ERROR_ESTIMATE.set(&[&sid], 0.0);
+            METRIC_CONTROL_OUTPUT.set(&[&sid], 0.0);
+        }
+
+        Self {
+            surb_production_estimator,
+            surb_consumption_estimator,
+            controller,
+            pid,
+            cfg,
+            session_id,
+            current_target_buffer: 0,
+            last_surbs_delivered: 0,
+            last_surbs_used: 0,
+            last_update: std::time::Instant::now(),
+            was_below_target: true,
+        }
+    }
+
+    /// Computes the next control update and adjusts the [`SurbFlowController`] rate accordingly.
+    #[tracing::instrument(level = "trace", skip(self), fields(session_id = %self.session_id))]
+    pub fn update(&mut self) -> u64 {
```

Comment:
Should this rather not be based on a threshold amount of received messages on that specific session?

Reply:
This works better imho, because the evaluation is continuous and reacts better to spikes.